### ISSUE TITLE
Crash Error: "NoClassDefFoundError"

### ISSUE THREAD

**Issue by daveharmon** [elastic/elasticsearch#10596]
Hi all, 
I'm working on getting started with elastic search (latest version).  All of my simple tests work great and the guide (which I have read several times and definitely completed every necessary step) is really helpful for general things.  However I plan on using elastic search as a dependency in a storm project so I need to deploy it to a single jar.  I've researched my error a lot and it seems like an issue with maven's shading that you guys use and I'm wondering if someone has some experience and can point me in the right direction.  I'm thinking that there is something wrong with my POM file that doesn't allow me to use certain functions I'm just not sure what to do at all.

The storm node successfully creates a node and client that connect to elastic search and this error comes when I attempt a client.prepareIndex().actionGet() call (specifics omitted).  The index and type that it tries to put into are already created on the elastic side (using curl).

My stack trace is something like this:
{ there are a bunch of storm errors here emanating from my elasticsearch storage bolt}
 Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.elasticsearch.index.codec.postingsformat.PostingFormats
     at org.elasticsearch.index.codec.CodecModule.configurePostingsFormats(CodecModule.java:126) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.index.codec.CodecModule.configure(CodecModule.java:178) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.AbstractModule.configure(AbstractModule.java:60) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.spi.Elements$RecordingBinder.install(Elements.java:204) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.spi.Elements.getElements(Elements.java:85) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.InjectorShell$Builder.build(InjectorShell.java:130) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:99) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.InjectorImpl.createChildInjector(InjectorImpl.java:131) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.inject.ModulesBuilder.createChildInjector(ModulesBuilder.java:69) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:328) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.cluster.metadata.MetaDataCreateIndexService$2.execute(MetaDataCreateIndexService.java:372) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:365) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:188) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:158) ~[mdec-storm-1.0-SNAPSHOT-jar-with-dependencies.jar:na]
     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) ~[na:1.7.0_75]
     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) ~[na:1.7.0_75]

and my pom file looks like this:
<project

    ...
     <parent>
         ...
     </parent>
     ...
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <skipTests>true</skipTests>
     </properties>

     <repositories>
         ...
     </repositories>

     <dependencies>
 ...
         
         <dependency>
             <groupId>org.elasticsearch</groupId>
             <artifactId>elasticsearch</artifactId>
         </dependency>
             
         ...
     </dependencies>

     <build>
         <sourceDirectory>src/main/java</sourceDirectory>
         <testSourceDirectory>src/test</testSourceDirectory>
         <resources>
             <resource>
                 <directory>src/main/resources</directory>
             </resource>
         </resources>
         <plugins>
             
             <plugin>
                 <artifactId>maven-assembly-plugin</artifactId>
                 <configuration>
                     <descriptorRefs>
                         <descriptorRef>jar-with-dependencies</descriptorRef>
                     </descriptorRefs>
                     <archive>
                         <manifest>
                             <mainClass></mainClass>
                         </manifest>
                     </archive>
                 </configuration>
                 <executions>
                     <execution>
                         <id>make-assembly</id>
                         <phase>package</phase>
                         <goals>
                             <goal>single</goal>
                         </goals>
                     </execution>
                 </executions>
             </plugin>
             
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-shade-plugin</artifactId>
                 <version>2.3</version>
                 <executions>
                     <execution>
                         <phase>package</phase>
                         <goals>
                             <goal>shade</goal>
                         </goals>
                     </execution>
                 </executions>
             </plugin>
             ...
         </plugins>
     </build>
 </project>

Thanks so much in advance

**Comment by dadoonet** [comment]
Does your final jar file contains the missing class? How does it look like?
Does it contain as well lucene classes we don't shade in Elasticsearch jar?
Try to add as a dependency as well all jars marked as optional in our pom.xml.

**Comment by daveharmon** [comment]
Thanks for your reply dadoonet,
I figured it out,  I had been using the maven assembly plugin to achieve a single jar as shown above.  Instead I decided to go with the maven shade plugin (v2.3) via this issue: https://github.com/elastic/elasticsearch/issues/3350 and add a series of AppendingTransformer transformers to the configuration section that combine those three specific META-INF/services files. 

Apparently these files get ignored in succeeding calls from the first, and elasticsearch requires all three, so I needed to use shading to append the similar files together

All good now, thanks for your help!
